### PR TITLE
Log1p Pressure Target: compress pressure dynamic range

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current pressure targets use an asinh transform (--asinh_pressure --asinh_scale 0.75). However, surface pressure distributions have extreme dynamic range: stagnation point pressures can be 10-100x larger than boundary layer values. MAE loss is dominated by these peak values, meaning the model optimizes stagnation/leading-edge accuracy at the expense of the rest of the surface.

**Log1p transform** — `sign(p) * log1p(|p|)` — compresses the dynamic range more aggressively than asinh for large values while preserving sign and being nearly linear for small values. This redistributes the loss contribution more evenly across the surface, which should improve mean surface MAE.

**Kaggle evidence:** Log-transforming skewed targets is one of the most reliable techniques in regression competitions. For heavy-tailed distributions, it consistently improves MAE/RMSE by 2-5% by reducing loss domination by outlier values.

**Key distinction from asinh:** asinh(x) ≈ ln(2|x|) for large x, so it already compresses. But log1p(|x|) compresses MORE aggressively: for x=100, asinh(100)≈5.3 while log1p(100)=4.6. The question is whether the additional compression helps or hurts precision on peak values.

**Important:** This changes the training target, NOT the evaluation metric. At inference, we invert the transform to get original-scale predictions. The evaluation must compute MAE on the original (untransformed) pressure values.

### Step 1: Add argument

```python
parser.add_argument('--log1p_pressure', action='store_true',
                    help='Apply sign(p)*log1p(|p|) transform to pressure targets')
```

### Step 2: Implement target transform

**This replaces the asinh transform for pressure, not in addition to it.**

When `--log1p_pressure` is set:
- Do NOT use `--asinh_pressure` (they are mutually exclusive)
- Apply to the pressure channel only (channel index 2 in the [Ux, Uy, p] target)

```python
def log1p_transform(p):
    """Sign-preserving log1p: sign(p) * log1p(|p|)"""
    return torch.sign(p) * torch.log1p(torch.abs(p))

def log1p_inverse(p_transformed):
    """Inverse: sign(p) * (exp(|p|) - 1)"""
    return torch.sign(p_transformed) * (torch.exp(torch.abs(p_transformed)) - 1)
```

Apply the transform to the pressure channel of targets after loading:
```python
if args.log1p_pressure:
    y[:, :, 2] = log1p_transform(y[:, :, 2])  # Transform pressure targets
```

### Step 3: Inverse transform at validation

At validation, the model predicts in log1p space. Invert before computing MAE:
```python
if args.log1p_pressure:
    pred[:, :, 2] = log1p_inverse(pred[:, :, 2])
    # Target is already in original space for MAE computation (or invert target too)
```

**Critical:** Ensure the MAE metric is computed on ORIGINAL-SCALE pressure, not transformed. The W&B logged metrics must be apples-to-apples with baseline.

### Step 4: Verify mutual exclusivity with asinh

Add a check:
```python
if args.log1p_pressure and args.asinh_pressure:
    raise ValueError("--log1p_pressure and --asinh_pressure are mutually exclusive")
```

### Step 5: Run 2 seeds (WITHOUT asinh_pressure)

```bash
cd cfd_tandemfoil && python train.py \
  --agent askeladd --wandb_name "askeladd/log1p-pressure-s42" --seed 42 \
  --wandb_group log1p-pressure-target \
  --log1p_pressure \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling

# Repeat for seed 73: --seed 73 --wandb_name "askeladd/log1p-pressure-s73"
```

**Note:** `--asinh_pressure` is intentionally REMOVED. The `--asinh_scale` flag is also not needed. All other flags remain identical.

**Debug check:** After a few epochs, verify that val metrics are being computed on original-scale pressure (not transformed). Compare the scale of reported p_in/p_oodc/p_tan/p_re to baseline values — if they're in a completely different range, the inverse transform isn't being applied correctly.

## Baseline

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in | 11.742 | < 11.742 |
| p_oodc | 7.643 | < 7.643 |
| p_tan | 27.874 | < 27.874 |
| p_re | 6.419 | < 6.419 |

- **Baseline W&B runs:** k5qwvce4 (seed 42), 7oa5xfhi (seed 73)
- **val/loss baseline:** ~0.37 (not directly comparable since target scale changes)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Log1p Pressure Target Transform

| Metric | Baseline (asinh) | Seed 42 (log1p) | Seed 73 (log1p) | Avg | Δ vs baseline |
|--------|-----------------|-----------------|-----------------|-----|---------------|
| **p_in** | 11.742 | 11.853 | 11.948 | 11.901 | **+1.4%** |
| **p_oodc** | 7.643 | 7.769 | **7.453** | 7.611 | **-0.4%** |
| **p_tan** | 27.874 | 28.550 | 29.099 | 28.825 | **+3.4%** |
| **p_re** | 6.419 | **6.393** | 6.411 | 6.402 | **-0.3%** |

### Run Details

- **Seed 42**: W&B run `ph59bpnw`, 147 epochs, ~75-81s/epoch, 46.1GB VRAM
- **Seed 73**: W&B run `05mq4e8z`, 146 epochs, ~73-82s/epoch, 46.4GB VRAM
- **Commands used**:
  ```
  CUDA_VISIBLE_DEVICES=0 python train.py --agent askeladd --seed 42 \
    --wandb_name "askeladd/log1p-pressure-s42" --wandb_group "log1p-pressure-target" \
    --log1p_pressure \
    --field_decoder --adaln_output --use_lion --lr 2e-4 \
    --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
    --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
    --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
    --pressure_first --pressure_deep \
    --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
    --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
    --gap_stagger_spatial_bias \
    --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
    --te_coord_frame --wake_deficit_feature --re_stratified_sampling
  ```
  (Same for seed 73 with `--seed 73`, `CUDA_VISIBLE_DEVICES=1`)

### Analysis

**Verdict: Negative — log1p does not improve over asinh baseline.**

The log1p pressure transform performs slightly worse than the asinh baseline on the key metrics:
- **p_in regresses +1.4%** (11.901 avg vs 11.742 baseline)
- **p_tan regresses +3.4%** (28.825 avg vs 27.874 baseline) — this is the biggest regression
- **p_oodc** and **p_re** are essentially flat (-0.4% and -0.3%, within noise)

**Why it didn't help:**

1. **Over-compression hurts peak accuracy**: The hypothesis predicted that log1p's more aggressive compression (log1p(100)=4.6 vs asinh(100)=5.3) would redistribute loss more evenly. But for surface pressure, peak values at stagnation/suction carry the most aerodynamic information. Compressing them too much loses gradient signal precisely where accuracy matters most — explaining the p_tan regression (tandem foils have extreme pressure gradients).

2. **asinh is already well-tuned**: The baseline uses `--asinh_scale 0.75`, which modulates the compression. asinh with scale<1 provides a gentler compression that preserves more peak-value discrimination than log1p. The asinh transform is effectively already in the sweet spot for this problem's dynamic range.

3. **No free lunch in target transforms**: The evaluation is on original-scale MAE, so the transform only affects training dynamics. Log1p shifts gradient focus toward the many low-pressure boundary-layer nodes, but these are already well-predicted. The net effect is slightly degraded accuracy on the harder stagnation/suction peak nodes.

### Suggested Follow-ups

1. **Scaled log1p**: Try `sign(p) * log1p(|p| * scale)` with scale < 1.0 (e.g., 0.5) to reduce compression strength, making it closer to asinh's behavior while preserving the log1p shape.
2. **Hybrid transform**: Apply log1p only to pressure values above a threshold (e.g., |p| > 50), keeping linear training signal for moderate pressures.
3. **Consider this idea closed**: The results suggest that asinh with tuned scale is already near-optimal for this problem's pressure distribution. Further target-transform experiments are unlikely to yield meaningful gains.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2329 because ADVISOR: Closing — log1p pressure target does not improve over asinh baseline. p_tan regresses +3.4%, p_in +1.4%, while p_oodc and p_re are within noise.

The existing asinh transform (with scale=0.75) is already well-tuned for our pressure distribution — it compresses large values while preserving small-value precision. Log1p compresses slightly more aggressively but doesn't improve the quality of the learned representation.

**Conclusion:** asinh_pressure remains the optimal target transform. No further target transform experiments needed. Moving you to a fresh experiment..
